src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.AnalyzerStateData.cs

```diff
@@ -52,13 +64,49 @@ public virtual void SetStateKind(StateKind stateKind)
             public void ResetToReadyState()
             {
                 SetStateKind(StateKind.ReadyToProcess);
+
+                if (_callbackForDeferredFreeOperation != null)
+                {
+                    _callbackForDeferredFreeOperation();
+                    _callbackForDeferredFreeOperation = null;
+                }
             }
 
             public virtual void Free()
             {
                 this.StateKind = StateKind.ReadyToProcess;
                 this.ProcessedActions.Clear();
             }
+
+            public void FreeAndReturnToPool<TAnalyzerStateData>(ObjectPool<TAnalyzerStateData> pool)
+                where TAnalyzerStateData : AnalyzerStateData
+            {
+                var state = (TAnalyzerStateData)this;
+                if (StateKind == StateKind.InProcess)
+                {
+                    // Do not free the state here if it is still being processed for analysis,
+                    // i.e. StateKind is 'InProcess'. Instead, marked it for a deferred free operation to
+                    // execute after the StateKind is reset to 'ReadyToProcess'.
+                    // This can happen in rare cases when multiple threads are trying to analyze the same
+                    // symbol/syntax/operation, such that one thread has finished all the analyzer callbacks,
+                    // but before this thread marks the state as complete and invokes this Free call,
+                    // second thread starts processing the same symbol/syntax/operation and sets the StateKind to
+                    // 'InProcess' for the same state object. If we free the state here in the first thread,
+                    // then this can lead to data corruption/exceptions in the second thread.
+                    // For example, see https://github.com/dotnet/roslyn/issues/59988.
+                    _callbackForDeferredFreeOperation = () => FreeAndReturnToPoolCore(state, pool);
+                }
+                else
+                {
+                    FreeAndReturnToPoolCore(state, pool);
```

Comment:
>FreeAndReturnToPoolCore(state, pool); [](http://example.com/codeflow?start=20&length=37)

Is there a guarantee that the other thread doesn't start processing the same symbol/syntax/operation once we get here?

In general, it feels like there is something wrong with the whole idea of sharing the same object between two different threads and the state moving "backwards" by one of them. Using an object pool for concurrently accessed objects also feels very suspicious.

Reply:
> Is there a guarantee that the other thread doesn't start processing the same symbol/syntax/operation once we get here

Yes. State object is guaranteed to have `StateKind.InProcess` when it is attempting to run analysis or access any properties off it.

> In general, it feels like there is something wrong with the whole idea of sharing the same object between two different threads and the state moving "backwards" by one of them

The state object can be fetched/read/written by only one thread at a time when attempting to execute analysis. The only exception is the bug being fixed here, i.e. when one of the threads has completed all analysis corresponding to the symbol/syntax/operation for the state and it needs to update the corresponding state map and subsequently free the state: https://github.com/dotnet/roslyn/blob/86e06cace23209576e17cb362b2501593c76cbd3/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.PerAnalyzerState.cs#L272-L287

Above code removes the state from the declaration data map, so the state would no longer be retrievable by future attempts. Attempt to free the state and return to pool should only happen when no other thread has hold of the state, which is what the above `StateKind.InProcess` check is ensuring.

for the code being added here which sets the callback for deferred free operation when the analysis corresponding to the state object has completed on the current thread and it invokes Free, but some other thread

Reply:
We discussed offline and decided that for such cases where we attempt to mark completion of state on one thread, such that another thread is still processing the state, we not bail out upfront and neither free the state nor mark it complete. This ensures that another thread will eventually re-attempt to mark completion of the state, at which time we will free the state and successfully mark completion. I have pushed a commit for this approach.